fv3gfs/util/utils.py

```diff
@@ -57,16 +59,44 @@ def ensure_contiguous(maybe_array: Union[np.ndarray, Storage]) -> None:
         raise ValueError("ndarray is not contiguous")
 
 
-def assign_array(
-    to_array: Union[np.ndarray, Storage], from_array: Union[np.ndarray, Storage],
+def safe_assign_array(
+    to_array: Union[np.ndarray, Storage], from_array: Union[np.ndarray, Storage]
 ):
-    # The cp.asarray call is required to explicitly copy the data
-    # in the case of numpy arrays or to prevent memory ownership
-    # errors on the cupy arrays from gt4py storages.
-    # This should be fixed in a later version of gt4py
-    if cp and isinstance(to_array, cp.ndarray):
-        to_array[:] = cp.asarray(from_array)
-    elif cp and isinstance(from_array, cp.ndarray):
-        to_array[:] = cp.asnumpy(from_array)
-    else:
+    """Failproof assignment for array on different devices.
+    
+    The memory will be downloaded/uploaded from GPU if need be.
+
+    Args:
+        to_array: destination ndarray
+        from_array: source ndarray
+    """
+    try:
         to_array[:] = from_array
+    except (ValueError, TypeError):
+        if cp and isinstance(to_array, cp.ndarray):
+            to_array[:] = cp.asarray(from_array)
+        elif cp and isinstance(from_array, cp.ndarray):
+            to_array[:] = cp.asnumpy(from_array)
+
+
+def device_synchronize(array: Union[np.ndarray, Storage]):
+    """Synchronize all memory communication"""
+    if cp and isinstance(array, cp.ndarray):
+        cp.cuda.runtime.deviceSynchronize()
+
+
+@contextlib.contextmanager
+def mpi_safe_allocator(allocator: Allocator):
+    """Make sure the allocator use complies with what MPI expect
+    
+    For G2G transfer, MPICH requires the allocation to not be done with managed
+    memory. Since we can't know what state `cupy` is in with switch for the default
+    pooled allocator.
+    """
+    if cp and allocator is cp.empty:
```

Comment:
We chatted about how you can refactor this into a function that returns the allocated array, and check whether we were given a cupy array in the else branch where it should really not be a cupy array.

Reply:
Refactored into a function. The bug is hard enough that I used a `raise RuntimeError` if a cupy allocation happens outside of the "safe" code path

Reply:
A `ValueError` (to say an argument had an invalid value) or `TypeError` (to say an argument didn't meet the type specifications) could be appropriate. RuntimeError is also fine.

Reply:
Actually could we use the `NotImplementedError` here? Technically we would need a `MissingImplementationError` since this not the dev fault but a lack in the internal logic